File: src/main/java/cs/ut/controller/TrainingController.java
package cs.ut.controller;

import com.google.common.collect.Lists;
import com.google.common.escape.Escaper;
import com.google.common.html.HtmlEscapers;
import cs.ut.config.MasterConfiguration;
import cs.ut.config.items.ModelParameter;
import cs.ut.config.items.Property;
import cs.ut.engine.JobManager;
import cs.ut.manager.LogManager;
import cs.ut.ui.FieldComponent;
import cs.ut.ui.NirdizatiGrid;
import cs.ut.ui.providers.GeneratorArgument;
import cs.ut.ui.providers.ModelParamToCombo;
import cs.ut.ui.providers.PropertyValueProvider;
import org.apache.log4j.Logger;
import org.zkoss.util.resource.Labels;
import org.zkoss.zk.ui.Component;
import org.zkoss.zk.ui.event.Events;
import org.zkoss.zk.ui.event.SelectEvent;
import org.zkoss.zk.ui.select.SelectorComposer;
import org.zkoss.zk.ui.select.annotation.Listen;
import org.zkoss.zk.ui.select.annotation.Wire;
import org.zkoss.zul.Button;
import org.zkoss.zul.Checkbox;
import org.zkoss.zul.Combobox;
import org.zkoss.zul.Comboitem;
import org.zkoss.zul.Vlayout;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class TrainingController extends SelectorComposer<Component> {
    private static final Logger log = Logger.getLogger(TrainingController.class);
    private static final String LEARNER = "learner";
    private static final String ENCODING = "encoding";
    private static final String BUCKETING = "bucketing";
    private static final String PREDICTION = "predictiontype";

    @Wire
    private Combobox clientLogs;

    @Wire
    private Combobox predictionType;

    @Wire
    private Checkbox advancedMode;

    @Wire
    private Button startTraining;

    @Wire
    private Vlayout gridContainer;

    private NirdizatiGrid<Property> parameterGrid;

    private transient Map<String, List<ModelParameter>> parameters = MasterConfiguration.getInstance().getModelConfiguration().getProperties();

    private transient List<Property> hyperParameters = new ArrayList<>();
    private transient List<Component> valueProviders = new ArrayList<>();

    @Override
    public void doAfterCompose(Component comp) throws Exception {
        super.doAfterCompose(comp);
        log.debug("Initialized TrainingController");
        parameterGrid = new NirdizatiGrid<>(new PropertyValueProvider());
        parameterGrid.setVisible(false);
        parameterGrid.setHflex("min");

        initClientLogs();
        initPredictions();
        initBasicMode();
    }

    private void initBasicMode() {
        gridContainer.getChildren().clear();
        NirdizatiGrid<GeneratorArgument> grid = new NirdizatiGrid<>(new ModelParamToCombo());
        grid.setHflex("min");

        grid.generate(
                parameters
                        .entrySet()
                        .stream()
                        .map(it -> new GeneratorArgument(it.getKey(), it.getValue()))
                        .collect(Collectors.toList()), true);

        valueProviders.addAll(grid.getFields().stream().map(FieldComponent::getControl).collect(Collectors.toList()));
        valueProviders.forEach(this::generateListener);

        gridContainer.appendChild(grid);
        gridContainer.appendChild(parameterGrid);
    }

    private void generateListener(Component control) {
        control.addEventListener(Events.ON_SELECT, e -> {
            log.debug(String.format("Combobox %s control changed, regenerating parameter grid.", control));
            log.debug(String.format("Old hyperparemeters are: %s", hyperParameters));
            hyperParameters.clear();
            hyperParameters.addAll(((ModelParameter) ((Comboitem) ((SelectEvent) e).getSelectedItems().iterator().next()).getValue()).getProperties());
            List<Component> components = valueProviders.stream().filter(it -> !it.equals(control)).collect(Collectors.toList());

            for (Component component : components) {
                Combobox combobox = (Combobox) component;
                List<Property> properties = ((ModelParameter) combobox.getSelectedItem().getValue()).getProperties();
                hyperParameters.addAll(properties);
            }
            log.debug(String.format("New hyperparameters: %s", hyperParameters));
            parameterGrid.generate(hyperParameters, true);

            if (!hyperParameters.isEmpty()) {
                parameterGrid.setVisible(true);
            }
        });
    }

    private void initPredictions() {
        List<ModelParameter> params = parameters.remove(PREDICTION);
        log.debug(String.format("Received %s prediction types", params.size()));

        params.forEach(it -> {
            Comboitem item = predictionType.appendItem(Labels.getLabel(it.getType().concat(".").concat(it.getId())));
            item.setValue(it);
        });

        predictionType.setSelectedItem(predictionType.getItemAtIndex(0));
        predictionType.setReadonly(true);
    }

    private void initClientLogs() {
        LogManager manager = LogManager.getInstance();
        List<File> fileNames = manager.getAllAvailableLogs();
        log.debug(String.format("Got %s items for client log combobox", fileNames.size()));

        Escaper escaper = HtmlEscapers.htmlEscaper();

        fileNames.forEach(file -> {
            Comboitem item = clientLogs.appendItem(escaper.escape(file.getName()));
            item.setValue(file);
        });

        if (!clientLogs.getItems().isEmpty()) {
            clientLogs.setSelectedItem(clientLogs.getItemAtIndex(0));
        } else {
            clientLogs.setDisabled(true);
        }
        clientLogs.setWidth("250px");
        clientLogs.setReadonly(true);

        clientLogs.addEventListener(Events.ON_CHANGE, e -> initBasicMode());
    }

    @Listen("onCheck = #advancedMode")
    public void enabledAdvanced() {
        if (advancedMode.isChecked()) {
            log.debug("enabling advanced mode");
//            initAdvancedMode();
        } else {
            log.debug("enabling basic mode");
            initBasicMode();
        }
    }

    @Listen("onClick = #startTraining")
    public void startTraining() {
        if (validateData()) {
            return;
        }
        log.debug("Parameters are valid, calling script to construct the model");
        Runnable jobs = () -> {
            JobManager.Manager.setLogFile(clientLogs.getSelectedItem().getValue());
            Comboitem comboitem = predictionType.getSelectedItem();

            Map<String, List<ModelParameter>> jobParameters = new HashMap<>();
            jobParameters.put(PREDICTION, Lists.newArrayList((ModelParameter)comboitem.getValue()));

            if (!advancedMode.isChecked()) {
                Component grid = gridContainer.getChildren().get(0);
                if (grid instanceof NirdizatiGrid) {
                    NirdizatiGrid g = (NirdizatiGrid) grid;
                    Map<String, Object> vals = g.gatherValues();

                    for (Map.Entry<String, Object> param : vals.entrySet()) {
                        String key = param.getKey();
                        ModelParameter val = new ModelParameter((ModelParameter) param.getValue());

                        if (LEARNER.equals(val.getType())) {
                                /* Set hyperparams */
                            List<Property> properties = parameterGrid.gatherValues()
                                    .entrySet()
                                    .stream()
                                    .map(it -> new Property(it.getKey(), "", it.getValue().toString()))
                                    .collect(Collectors.toList());
                            val.getProperties().clear();
                            val.setProperties(properties);
                        }
                        jobParameters.put(key, Lists.newArrayList(val));
                    }
                }
                JobManager.Manager.generateJobs(jobParameters);
            }

            JobManager.Manager.deployJobs();
        };

        log.debug("Jobs generated...");
        jobs.run();
    }

    private boolean validateData() {
        if (advancedMode.isChecked()) {
            return true;
        } else {
            boolean valid = true;
            for (Component comp : gridContainer.getChildren()) {
                if (comp instanceof NirdizatiGrid && ((NirdizatiGrid) comp).validate()) {
                    valid = false;
                }
            }
            return valid;
        }
    }
}

File: src/main/kotlin/cs/ut/controllers/TrainingController.kt
package cs.ut.controllers

import com.google.common.html.HtmlEscapers
import cs.ut.config.MasterConfiguration
import cs.ut.config.items.ModelParameter
import cs.ut.controllers.training.AdvancedModeController
import cs.ut.controllers.training.BasicModeController
import cs.ut.controllers.training.ModeController
import cs.ut.engine.JobManager
import cs.ut.engine.LogManager
import cs.ut.util.NirdizatiUtil
import cs.ut.util.OUTCOME
import cs.ut.util.readLogColumns
import org.apache.commons.io.FilenameUtils
import org.apache.log4j.Logger
import org.zkoss.util.resource.Labels
import org.zkoss.zk.ui.Component
import org.zkoss.zk.ui.Executions
import org.zkoss.zk.ui.event.Events
import org.zkoss.zk.ui.event.SelectEvent
import org.zkoss.zk.ui.select.SelectorComposer
import org.zkoss.zk.ui.select.annotation.Listen
import org.zkoss.zk.ui.select.annotation.Wire
import org.zkoss.zul.*
import java.io.File

class TrainingController : SelectorComposer<Component>(), Redirectable {
    private val log: Logger = Logger.getLogger(TrainingController::class.java)!!

    companion object {
        const val LEARNER = "learner"
        const val ENCODING = "encoding"
        const val BUCKETING = "bucketing"
        const val PREDICTION = "predictiontype"

        val DEFAULT = MasterConfiguration.defaultValuesConfiguration.minValue
        val AVERAGE = MasterConfiguration.defaultValuesConfiguration.average.toString()
    }

    @Wire
    private lateinit var clientLogs: Combobox

    @Wire
    private lateinit var predictionType: Combobox

    @Wire
    private lateinit var advancedMode: Checkbox

    @Wire
    private lateinit var gridContainer: Vlayout

    @Wire
    private lateinit var thresholdContainer: Hbox

    private lateinit var radioGroup: Radiogroup

    private lateinit var gridController: ModeController

    override fun doAfterCompose(comp: Component?) {
        super.doAfterCompose(comp)

        radioGroup = Radiogroup()

        initClientLogs()
        initPredictions()

        gridController = BasicModeController(gridContainer, getLogFileName())
    }

    private fun getLogFileName(): String = FilenameUtils.getBaseName((clientLogs.selectedItem.getValue() as File).name)

    private fun initPredictions() {
        predictionType.items.clear()
        log.debug("Cleared prediction type items")

        val params: List<ModelParameter> = MasterConfiguration.modelConfiguration.properties[PREDICTION]!!
        log.debug("Received ${params.size} prediciton types")

        val logFile: File = clientLogs.selectedItem.getValue<File>() ?: return

        val dataSetColumns: List<String> = readLogColumns(FilenameUtils.getBaseName(logFile.name))

        params.forEach {
            val item: Comboitem = predictionType.appendItem(NirdizatiUtil.localizeText("${it.type}.${it.id}"))
            val modelParam = ModelParameter(it)
            item.setValue(modelParam)

            if (modelParam.id == OUTCOME) {
                modelParam.setUpRadioButtons()
            }
        }

        dataSetColumns.forEach {
            val modelParameter = ModelParameter(it, it, PREDICTION, true, mutableListOf())
            modelParameter.translate = false
            val item: Comboitem = predictionType.appendItem(modelParameter.id)
            item.setValue(modelParameter)
        }

        predictionType.addEventListener(Events.ON_SELECT, { e ->
            e as SelectEvent<*, *>
            val param = (e.selectedItems.first() as Comboitem).getValue() as ModelParameter
            log.debug("Prediction type model changed to $param")
            if (param.id == OUTCOME) {
                thresholdContainer.isVisible = true
                log.debug("Prediciton type is $OUTCOME generating radio buttons")
            } else {
                log.debug("Clearing thresholdContainer")
                thresholdContainer.isVisible = false
            }
        })

        predictionType.selectedItem = predictionType.items[0]
        predictionType.isReadonly = true
    }

    private fun ModelParameter.setUpRadioButtons() {
        val avg = radioGroup.appendItem(NirdizatiUtil.localizeText("threshold.avg"), this.parameter)
        avg.setValue(this.parameter.toDouble())
        radioGroup.selectedItem = avg

        val custom = radioGroup.appendItem(NirdizatiUtil.localizeText("threshold.custom"), this.parameter)
        custom.setValue(DEFAULT)
        val customBox = Doublebox()
        customBox.width = "60px"
        customBox.setValue(DEFAULT)
        customBox.style = "padding-top: 10px"
        customBox.vflex = "1"
        customBox.addEventListener(Events.ON_CHANGE, { _ ->
            log.debug("New value for custom threshold ${customBox.value}")
            val res: Double? = customBox.value

            if (res == null || res <= 0) {
                custom.setValue(DEFAULT)
                customBox.setValue(DEFAULT)
                customBox.errorMessage = NirdizatiUtil.localizeText("threshold.custom_error", 0)
            } else {
                customBox.clearErrorMessage()
                custom.setValue(res)
            }
        })

        thresholdContainer.appendChild(radioGroup)
        thresholdContainer.appendChild(customBox)
    }

    private fun initClientLogs() {
        val files = LogManager.getAllAvailableLogs()
        log.debug("Found ${files.size} log files")

        val escaper = HtmlEscapers.htmlEscaper()
        files.forEach {
            val item = clientLogs.appendItem(escaper.escape(it.name))
            item.setValue(it)
        }

        if (clientLogs.itemCount > 0) {
            clientLogs.selectedItem = clientLogs.items[0]
        } else {
            clientLogs.isDisabled = true
        }

        clientLogs.isReadonly = true

        clientLogs.addEventListener(Events.ON_SELECT, { _ ->
            switchMode()
            initPredictions()
        })
    }

    @Listen("onCheck = #advancedMode")
    fun switchMode() {
        gridController = when (advancedMode.isChecked) {
            true -> AdvancedModeController(gridContainer)
            false -> BasicModeController(gridContainer, getLogFileName())
        }
    }

    @Listen("onClick = #startTraining")
    fun startTraining() {
        if (!gridController.isValid()) return

        val jobParamters = mutableMapOf<String, List<ModelParameter>>()
        jobParamters.put(PREDICTION, listOf(predictionType.selectedItem.getValue()))
        jobParamters.putAll(gridController.gatherValues())

        if (!jobParamters.validateParameters()) return

        val prediction: ModelParameter = jobParamters[PREDICTION]!!.first()

        if (prediction.id == OUTCOME) {
            val value = (radioGroup.selectedItem.getValue() as Double)
            prediction.parameter = if (value == -1.0) AVERAGE else value.toString()
        }

        log.debug("Parameters are valid, calling script to train the model")
        val jobThread = Runnable {
            JobManager.logFile = clientLogs.selectedItem.getValue()
            JobManager.generateJobs(jobParamters)
            JobManager.deployJobs()
        }
        jobThread.run()
        log.debug("Job generation thread started")
    }

    private fun Map<String, List<ModelParameter>>.validateParameters(): Boolean {
        var isValid = true
        var msg = ""

        if (this[ENCODING] == null) {
            msg += ENCODING
            isValid = false
        }

        if (this[BUCKETING] == null) {
            msg += if (msg == "") BUCKETING else ", $BUCKETING"
            isValid = false
        }

        if (this[LEARNER] == null) {
            msg += if (msg == "") LEARNER else ", $LEARNER"
            isValid = false
        }

        if (!isValid) {
            NirdizatiUtil.showNotificationAsync(
                    Labels.getLabel("training.validation_failed", arrayOf(msg)),
                    Executions.getCurrent().desktop,
                    "error"
            )
        }

        return isValid
    }
}